### ISSUE TITLE
Docker Image CVEs

### ISSUE THREAD

**Issue by tedpatrick** [sartography/spiff-arena#547]
In reviewing the backend docker containers for SpiffArena, there are many CVEs. These accumulate in time on the base Docker image layers. 

## Free scan via ECR: 
- Set the ECR repository to `scan-on-push`
```
docker pull ghcr.io/sartography/spiffworkflow-backend:latest
docker tag ghcr.io/sartography/spiffworkflow-backend:latest {account}.dkr.ecr.{region}.amazonaws.com/{ecr_repo}:swb_latest
docker push {account}.dkr.ecr.{region}.amazonaws.com/{ecr_repo}:swb_latest
```

<img width="1562" alt="Screenshot 2023-10-13 at 10 40 46 AM" src="https://github.com/sartography/spiff-arena/assets/35942272/46f88f26-8424-4ec7-a7a1-90931aae1cfe">

This will block our use of Spiff and in the spirit of transparency, I want to make maintainers aware.

**Comment by burnettk** [comment]
@tedpatrick , thank you very much for the report. we definitely don't want to see security concerns stopping anyone from using Spiff. We take these issues very seriously. We've been lucky to have had the opportunity for multiple third parties to review the source code, and we have mitigated all concerns. We do run snyk in CI, https://github.com/sartography/spiff-arena/actions/runs/6509463582, but there is an obvious gap here in our security scanning automation, and we will look to close that up rapidly. In the meantime, the base image has been updated, and the resulting image has 322 fewer vulnerabilities. There is one critical vulnerability, and no high severity issues, according to docker scan. The critical vulnerability, https://security.snyk.io/vuln/SNYK-DEBIAN12-NGHTTP2-5953379, was earned through `apt-get install curl` and unfortunately "There is no fixed version for Debian:12."

The docker images in question are:

 * https://github.com/sartography/docker-python/blob/main/Dockerfile aka `ghcr.io/sartography/python:3.11`
 * https://github.com/sartography/spiff-arena/blob/main/spiffworkflow-backend/Dockerfile aka `ghcr.io/sartography/spiffworkflow-backend:latest` / `ghcr.io/sartography/spiffworkflow-backend:main-2023-10-14_03-32-24` / `ghcr.io/sartography/spiffworkflow-backend:v0.0.45`

If you see any opportunity to improve the Dockerfiles, or if the image is still not meeting expectations, please let us know. If it would be easier to discuss synchronously, please reach out via our website or discord.

I'm going to leave this issue open until checks have been added to our CI process.

Thank you again for bringing this to our attention. :bow:

**Comment by tedpatrick** [comment]
Thanks for the great response. This is one of those hard ongoing problems.

The base python docker images are a great base but we found we got more control building it up ourselves.

```
# builder image.
FROM {AWS_ACC}.dkr.ecr.us-east-1.amazonaws.com/server:builder_1 as builder
COPY apps/server/api/locks/environment-linux-64.txt /locks/environment-linux-64.txt
RUN conda-lock install -p /opt/env --copy /locks/environment-linux-64.txt

# platform image.
FROM --platform=linux/amd64 debian:bookworm-slim
COPY --from=builder /opt/env /opt/env

ARG APP_DIR
ARG SHARED_DIR

ENV PATH="/opt/env/bin:${PATH}"

WORKDIR /app

# Copy the shared and application code
COPY ${SHARED_DIR} /app/shared
COPY ${APP_DIR} /app

CMD ["uvicorn", "server.app:app", "--host", "0.0.0.0", "--port", "80", "--proxy-headers"]
```

This one used a builder with conda-lock to build a python+deps from conda-forge, then the environment is copied into a clean base image using debian:bookworm-slim. The builder allows you to keep any dev tools out of the production image. 

The builder Dockerfile:
```
FROM continuumio/miniconda:latest as builder
COPY /locks/environment-linux-64.txt /locks/environment-linux-64.txt
RUN conda create -p /opt/builder --copy --file /locks/environment-linux-64.txt
ENV PATH="/opt/builder/bin:${PATH}"
```

Again, this is a hard problem, great to see a proactive posture around security.

Ted :)

**Comment by tedpatrick** [comment]
Updated Scan. Looks to be mostly system tools with outstanding CVEs now.


<img width="1403" alt="Screenshot 2023-10-14 at 10 32 06 AM" src="https://github.com/sartography/spiff-arena/assets/35942272/c5bf8b91-1635-48e3-8a06-409f67bb0af9">

**Comment by burnettk** [comment]
We have overhauled our security scanning that occurs at build time. We added a pile of snyk actions that are now scanning both frontend and backend, docker images, libraries, and code. Previously we had been scanning just backend libraries. A number of issues were surfaced, and we have fixed a number of them. We will continue to make progress on this over time, but the framework now exists. Thank you very much, @tedpatrick ! Are you unblocked? :)
